docs/notebooks/autodiff_remat.md

```diff
@@ -0,0 +1,548 @@
+
+By applying `jax.checkpoint` in the right way, there are many tradeoffs between memory usage and (re)computation that can be expressed. One surprising example is _recursive_ checkpointing, where we apply `jax.checkpoint` to a function which itself calls `jax.checkpoint`-decorated functions in a way so that memory usage from the chain composition of $D$ functions scales like $\mathcal{O}(\log_2 D)$ rather than $\mathcal{O}(D)$.
+
+As a toy example, consider the chain composition of multiple `jnp.sin` functions:
+
+```{code-cell}
+def chain_compose(funs):
+  def f(x):
+    for fun in funs:
+      x = fun(x)
+    return x
+  return f
+
+f = chain_compose([jnp.sin] * 8)
+print_saved_residuals(f, 3.)
+```
+
++++ {"id": "3SFXo4n6YJQG"}
+
+In general, the number of stored residuals scales linearly with the length of the chain:
+
+```{code-cell}
+f = chain_compose([jnp.sin] * 16)
+print_saved_residuals(f, 3.)
+```
+
++++ {"id": "RcTuFRwZYXm7"}
+
+But we can apply `jax.checkpoint` recursively to improve the scaling:
+
+```{code-cell}
+def recursive_checkpoint(funs):
+  if len(funs) == 1:
+    return funs[0]
+  elif len(funs) == 2:
+    f1, f2 = funs
+    return lambda x: f1(f2(x))
+  else:
+    f1 = recursive_checkpoint(funs[:len(funs)//2])
+    f2 = recursive_checkpoint(funs[len(funs)//2:])
+    return lambda x: f1(jax.checkpoint(f2)(x))
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_saved_residuals(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 16)
+print_saved_residuals(f, 3.)
+```
+
++++ {"id": "D0yhcndHX0yN"}
+
+The cost here, as usual, is recomputation: in particular, we end up performing $\mathcal{O}(\log_2 D)$ times as many FLOPs:
+
+```{code-cell}
+f = chain_compose([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
+```{code-cell}
+f = recursive_checkpoint([jnp.sin] * 8)
+print_fwd_bwd(f, 3.)
+```
+
++++ {"id": "wvSm1yu0yUtd"}
+
+### Practical notes
+
++++ {"id": "LSADkBOCyX9X"}
+
+When differentiated functions are staged out to XLA for compilation, for example by applying `jax.jit` to a function which contains a `jax.grad` call, XLA will automatically optimize the computation, including decisions about when to compute or rematerialize values. As a result, **`jax.checkpoint` often isn't needed for differentiated functions under a `jax.jit`**. XLA will optimize things for you.
+
+One exception is when using staged-out control flow, like `jax.lax.scan`. Automatic compiler optimizations between multiple control flow primitives, e.g. between a forward-pass `scan` and the corresponding backward-pass `scan`, aren't aren't as thorough. As a result, it's often a good idea to use `jax.checkpoint` on the body function we pass to `jax.lax.scan`.
+
+For example, one common pattern in large Transformer models is to express the architecture as a `jax.lax.scan` over layers so as to reduce compilation times. That is, using a simple fully-connected network as an analogy, instead of writing something like this:
```

Comment:
```suggestion
For example, one common pattern in large [Transformer](https://arxiv.org/abs/1706.03762) models is to express the architecture as a `jax.lax.scan` over layers so as to reduce compilation times. To use a simple fully-connected network as an analogy, instead of writing something like this:
```
or, alternatively, link to the [Wikipedia article](https://en.wikipedia.org/wiki/Transformer_(machine_learning_model))?

Reply:
Done!